Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -1,443 +1,381 @@
+        // Run the actual operation
+        [operation start];
+    });
 }
-@end
 
-@implementation NSOperationQueue
-static void ClearList(NSAtomicListRef* listPtr) {
-    for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
-        while (NSAtomicListPop(&listPtr[i])) {
-            ;
+// Private helper that starts as many queued operations as possible
+- (void)_popAndStartQueuedOperations {
+    if (!self.suspended) {
+        std::lock_guard<std::recursive_mutex> lock(_concurrentOperationCountLock);
+        while (([self _belowMaxConcurrentOperations]) && (!_queuedOperations.empty())) {
+            NSOperation* operation = _queuedOperations.front();
+            _queuedOperations.pop();
+            [self _startOperation:operation];
         }
     }
 }
 
-static BOOL RunOperationFromLists(NSAtomicListRef* listPtr, NSAtomicListRef* sourceListPtr, id* curOperation) {
-    StrongId<NSOperation> op = NSAtomicListPop(listPtr);
-    if (op == nil) {
-        *listPtr = NSAtomicListSteal(sourceListPtr);
-        // source lists are in LIFO order, but we want to execute operations in the order they were enqueued
-        // so we reverse the list before we do anything with it
-        NSAtomicListReverse(listPtr);
-        op = NSAtomicListPop(listPtr);
-    }
+// Private helper that sets _dispatchQueue based on the current quality of service
+- (void)_setDispatchQueueUsingQualityOfService {
+    std::lock_guard<std::recursive_mutex> lock(_dispatchQueueLock);
+    _dispatchQueue = dispatch_get_global_queue(_QOSClassForNSQualityOfService(self.qualityOfService), 0);
+}
 
-    if (op != nil) {
-        if ([op isReady]) {
-            *curOperation = op;
-            [op start];
-            *curOperation = nil;
-        } else {
-            NSAtomicListInsert(sourceListPtr, op);
-            // We claim that we have done some work here.
-            // Otherwise, this work session is considered to be over and we clear the pending queue.
-            return TRUE;
-        }
-    }
++ (BOOL)automaticallyNotifiesObserversOfOperations {
+    // This class manually notifies for operations, which generates changes in addOperation/s.
+    return NO;
+}
 
-    return op != nil;
++ (BOOL)automaticallyNotifiesObserversOfOperationCount {
+    // This class manually notifies for operationCount, as it changes at the same time as operations
+    return NO;
 }
 
-- (id)_workThread {
-    BOOL stop = FALSE;
+- (instancetype)init {
+    if (self = [super init]) {
+        _operations.attach([NSMutableArray new]);
+        _maxConcurrentOperationCount = NSOperationQueueDefaultMaxConcurrentOperationCount;
+        _qualityOfService = NSQualityOfServiceDefault;
 
-    while (!stop) {
-        [self _doMainWork];
-        pthread_mutex_lock(&priv->_threadRunningLock);
-        if (![self hasMoreWork]) {
-            stop = TRUE;
-            priv->_thread = nil;
+        if (dispatch_queue_t underlyingQueue = [self underlyingQueue]) {
```

Comment:
! ! !

This wasn't allowed before (C++11); do we have CC on whether we can do this?